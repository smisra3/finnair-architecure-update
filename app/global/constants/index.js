/* TODO: Remove lint exception */
/* eslint-disable import/prefer-default-export */
export const DEVICE_TYPE = 'DEVICE_TYPE';
export const IS_TABLET = 'IS_TABLET';
export const REGION_TYPE = 'REGION_TYPE';
export const CONFIG_KEYS = 'CONFIG_KEYS';
export const CURRENT_ROUTE = 'CURRENT_ROUTE';
export const PAGE_URL = 'PAGE_URL';
export const PAGE_QUERY = 'PAGE_QUERY';
export const GLOBAL_DATA_FAILURE = 'FIL/GLOBAL/GLOBAL_DATA_FAILURE';
export const CHECK_AUTH = 'FIL/GLOBAL/CHECK_AUTH';
export const CHECK_AUTH_SUCCESS = 'FIL/GLOBAL/CHECK_AUTH_SUCCESS';
export const SET_USER_STATE = 'FIL/GLOBAL/SET_USER_STATE';
export const PAGE_ORIGIN = 'PAGE_ORIGIN';
export const LABELS_STATE_KEY = 'labels';

export const USER_STATE_LOGGED_IN = 'FIL/GLOBAL/SUCCESS';
export const USER_STATE_RECOGNIZED = 'FIL/GLOBAL/RECOGNIZED';
export const USER_STATE_ANONYMOUS = 'FIL/GLOBAL/ANONYMOUS';

export const GET_LABELS = 'FIL/GLOBAL/GET_LABELS';
export const GET_LABELS_SUCCESS = 'FIL/GLOBAL/GET_LABELS_SUCCESS';
export const GET_LABELS_FAILURE = 'FIL/GLOBAL/GET_LABELS_FAILURE';
