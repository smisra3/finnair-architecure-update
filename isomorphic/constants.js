const HEARTBEAT_SESSION_NAME = 'JSESSIONID';
const ENV_DEVELOPMENT = 'development';
const ENV_PRODUCTION = 'production';
const SERVICE_WORKER_FILE = '/service-worker.js';
const API_PROXY_PATH = '/api';
const CLEAR_SERVICES_CACHE = '/clear-services-cache';
const HEALTH_CHECK = '/healthcheck';
const PROTOCOL_HTTPS = 'https';
const RESPONSE_OK = 200;
const RESPONSE_INTERNAL_SERVER_ERROR = 500;
const STATIC_FOLDER_PREFIX = '/static/';

module.exports = {
  HEARTBEAT_SESSION_NAME,
  ENV_DEVELOPMENT,
  ENV_PRODUCTION,
  SERVICE_WORKER_FILE,
  API_PROXY_PATH,
  CLEAR_SERVICES_CACHE,
  PROTOCOL_HTTPS,
  RESPONSE_OK,
  RESPONSE_INTERNAL_SERVER_ERROR,
  STATIC_FOLDER_PREFIX,
  HEALTH_CHECK,
};
