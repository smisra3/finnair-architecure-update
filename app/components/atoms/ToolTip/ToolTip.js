/**
 *
 * ToolTip
 *
 */
// @flow
import { Component } from 'react';
import type { Node } from 'react';
import styled from 'styled-components';
import styles from './ToolTip.style';
import { MOBILE, KEY_ENTER, KEY_SPACE } from '../../../constants';
import { labelCloseToolTip } from '../../../locales/en-US';

type Props = {
  children: Node,
  className: string,
  label: string,
  container: string,
  styleOverride: ?Object,
  onClick?: () => void,
  deviceType?: string,
  isCheckoutPage: boolean,
  isShoppingBag: boolean,
  isLargeToolTip?: boolean,
  defaultMobileToolTip: boolean,
  excludeContainer: boolean,
};

type State = {
  isVisible: boolean,
  style: Object,
};

class ToolTip extends Component<Props, State> {
  static defaultProps = {
    isLargeToolTip: false,
    onClick: () => {},
    deviceType: '',
  };

  state = {
    isVisible: false,
    style: {
      width: this.props.deviceType === MOBILE ? 'calc(100vw - 30px)' : '13em',
      left: '-100%',
    },
  };

  componentDidMount() {
    this.setLabelPosition(
      this.props.container,
      this.props.styleOverride,
      this.props.excludeContainer,
    );
  }

  componentWillReceiveProps(nextProps: Props) {
    const {
      deviceType,
      isCheckoutPage,
      isShoppingBag,
      defaultMobileToolTip,
      excludeContainer,
    } = this.props;
    const isMobile = deviceType === MOBILE;
    const leftPosition = `-${this.elementRef.offsetLeft - 15}px`;
    if (defaultMobileToolTip && isMobile) {
      const defaultMobileStyle = {
        width: 'calc(100vw - 30px)',
        left: leftPosition,
      };
      this.setLabelPosition(nextProps.container, defaultMobileStyle, excludeContainer);
    }
    if (
      this.props.styleOverride &&
      nextProps.styleOverride.width !== this.props.styleOverride.width &&
      deviceType === MOBILE &&
      (isCheckoutPage || isShoppingBag)
    ) {
      const styleOverride = {
        width: nextProps.styleOverride.width,
        left: `-${this.elementRef.offsetLeft - 15}px`,
      };
      this.setLabelPosition(nextProps.container, styleOverride, excludeContainer);
    }
  }

  setLabelPosition(container, styleOverride, excludeContainer) {
    const style = {};
    const { deviceType } = this.props;
    const isMobile = deviceType === MOBILE;
    let substractOffset = 30;

    if (isMobile && excludeContainer) {
      substractOffset = 0;
    }

    if (container && this.elementRef) {
      const containerElement = document.querySelector(`[data-tool-tip="${container}"]`);
      let toolTipWidth;
      if (!styleOverride && containerElement) {
        toolTipWidth = isMobile
          ? containerElement.offsetWidth - substractOffset
          : containerElement.offsetWidth;
      }
      if (containerElement && this.elementRef) {
        style.width = styleOverride ? `${styleOverride.width}` : `${toolTipWidth}px`;
        style.left = styleOverride
          ? `${styleOverride.left}`
          : `${containerElement.offsetLeft - this.elementRef.offsetLeft}px `;
      }
      this.setState({ style });
    }
  }

  elementRef = null;

  handleMouseEnter(e: SyntheticEvent<>, deviceType: string) {
    if (deviceType !== MOBILE) {
      this.setState({
        isVisible: true,
      });
    }
  }

  handleMouseLeave() {
    this.setState({
      isVisible: false,
    });
  }

  handleKeyPress(event: SyntheticEvent<>) {
    if (event.which === KEY_SPACE || event.which === KEY_ENTER) {
      this.setState({
        isVisible: !this.state.isVisible,
      });
    }
  }

  showVisibility = () => {
    this.setState({
      isVisible: true,
    });
  };

  hideVisibility = (e) => {
    const tooltipLinkExist = this.elementRef.getElementsByClassName('tooltip-link').length;
    if (!tooltipLinkExist || e.target.className.indexOf('tooltip-link') !== -1) {
      this.setState({
        isVisible: false,
      });
    }
  };

  handleClick() {
    this.setState(
      {
        isVisible: !this.state.isVisible,
      },
      () => {
        if (this.props.deviceType === MOBILE) {
          const defaultMobileStyle = {
            width: 'calc(100vw - 30px)',
            left: `-${this.elementRef.offsetLeft}px`,
          };
          this.setLabelPosition(
            this.props.container,
            defaultMobileStyle,
            this.props.excludeContainer,
          );
        }
      },
    );
    if (this.props.onClick) {
      this.props.onClick(this.state.isVisible);
    }
  }

  render() {
    const isVisible = this.state.isVisible ? ' is-tooltip-visible' : ' is-tooltip-hidden';
    const className = `${this.props.className} tooltip-parent ${isVisible}`;
    const { deviceType, isLargeToolTip } = this.props;
    return (
      <button
        className={className}
        onFocus={deviceType !== MOBILE ? () => this.showVisibility() : null}
        onBlur={deviceType !== MOBILE ? e => this.hideVisibility(e) : null}
        onKeyPress={e => this.handleKeyPress(e)}
        onMouseEnter={e => this.handleMouseEnter(e, deviceType)}
        onMouseLeave={() => this.handleMouseLeave()}
        ref={(element) => {
          this.elementRef = element;
        }}
        onClick={() => {
          this.handleClick();
        }}
      >
        <span
          role="presentation"
          data-slnm-id="toolTipCTA"
          className="tooltip-icon"
          onClick={() => {
            this.handleClick();
          }}
        >
          {this.props.children}
        </span>
        <span className="tooltip-label" data-slnm-id="toolTipBubble" style={this.state.style}>
          {isLargeToolTip ? (
            <span className="tooltip-label-child large-tooltip  tooltip-link">
              <i className="icon close" role="button" aria-label={labelCloseToolTip} />
              {this.props.label}
            </span>
          ) : (
            <span className="tooltip-label-child">{this.props.label}</span>
          )}
        </span>
      </button>
    );
  }
}
export default styled(ToolTip)`${styles}`;

export { ToolTip };
