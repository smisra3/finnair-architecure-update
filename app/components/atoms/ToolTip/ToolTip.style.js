import { css } from 'styled-components';
import { BRAND_THEME, PRIMARY_THEME, DESKTOP, MOBILE } from '../../../constants';

/* stylelint-disable no-descending-specificity */
export const commonStyles = css`
  position: relative;
  border: none;

  .tooltip-icon {
    &::after,
    &::before {
      left: 50%;
      border: 10px solid transparent;
      content: '';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &::after {
      border-color: transparent;
      border-width: 10px;
      margin-left: -10px;
      border-top-color: ${props => props.theme.primaryBgColor};
      z-index: 3;
    }

    &::before {
      border-color: transparent;
      border-width: 11px;
      margin-left: -11px;
      border-top-color: ${props => props.theme.secondaryBgColor};
      z-index: 1;
    }
  }

  .tooltip-label {
    z-index: 2;
    position: absolute;
    padding-bottom: 1em;
    bottom: 100%;

    .tooltip-label-child {
      display: block;
      border-radius: 2px;
      padding: 0.5em;
      background-color: ${props => props.theme.primaryBgColor};
      border: 1px solid ${props => props.theme.secondaryBgColor};
      font-family: ${props => props.theme.secondaryFontFamily};

      button {
        font-family: ${props => props.theme.secondaryFontFamily};
        text-align: left;
        line-height: 15px;
      }
    }

    a {
      color: #000;
    }
  }

  &.is-tooltip-visible {
    .tooltip-icon {
      &::after,
      &::before {
        opacity: 1;
        visibility: visible;
      }
    }

    .tooltip-label {
      opacity: 1;
      visibility: visible;
    }
  }

  &.is-tooltip-hidden {
    .tooltip-icon {
      &::after,
      &::before {
        opacity: 0;
        visibility: hidden;
      }
    }

    .tooltip-label {
      opacity: 0;
      visibility: hidden;
    }
  }
`;

export const loftStyles = css`
  .tooltip-icon {
    &::after,
    &::before {
      top: ${props => props.top || '-100% '};
    }
  }
`;

export const factoryStyles = css`
  .tooltip-icon {
    &::after,
    &::before {
      top: ${props => props.top || '-81% '};
    }
  }
`;

export const commonMobileStyles = css`
  .tooltip-label-child.large-tooltip {
    padding: 12px;
    position: relative;
    font-size: 12px;

    .close {
      position: absolute;
      right: 5px;
      top: 5px;
    }
  }
`;
/* TODO: refactor inherited styles implementation for a better solution */
export default {
  [BRAND_THEME]: {
    [DESKTOP]: css`
      ${commonStyles};
      ${loftStyles};
      ${props =>
    (props.inheritedStyles && props.inheritedStyles[BRAND_THEME]
      ? props.inheritedStyles[BRAND_THEME][DESKTOP]
      : '')};
      ${props =>
    (props.inheritedStyles && !props.inheritedStyles[BRAND_THEME]
      ? props.inheritedStyles[DESKTOP]
      : '')};
      ${props =>
    (props.inheritedStyles &&
        !props.inheritedStyles[BRAND_THEME] &&
        !props.inheritedStyles[DESKTOP]
      ? props.inheritedStyles
      : '')};
    `,
    [MOBILE]: css`
      ${commonStyles};
      ${loftStyles};
      ${commonMobileStyles};
      ${props =>
    (props.inheritedStyles && props.inheritedStyles[BRAND_THEME]
      ? props.inheritedStyles[BRAND_THEME][MOBILE]
      : '')};
      ${props =>
    (props.inheritedStyles && !props.inheritedStyles[BRAND_THEME]
      ? props.inheritedStyles[MOBILE]
      : '')};
      ${props =>
    (props.inheritedStyles &&
        !props.inheritedStyles[BRAND_THEME] &&
        !props.inheritedStyles[MOBILE]
      ? props.inheritedStyles
      : '')};

      span.tooltip-label-child.large-tooltip {
        font-size: 15px;
        line-height: 1.15;
      }
    `,
  },
  [PRIMARY_THEME]: {
    [DESKTOP]: css`
      ${commonStyles};
      ${factoryStyles};
      ${props =>
    (props.inheritedStyles && props.inheritedStyles[PRIMARY_THEME]
      ? props.inheritedStyles[PRIMARY_THEME][DESKTOP]
      : '')};
      ${props =>
    (props.inheritedStyles && !props.inheritedStyles[PRIMARY_THEME]
      ? props.inheritedStyles[DESKTOP]
      : '')};
      ${props =>
    (props.inheritedStyles &&
        !props.inheritedStyles[PRIMARY_THEME] &&
        !props.inheritedStyles[DESKTOP]
      ? props.inheritedStyles
      : '')};
    `,
    [MOBILE]: css`
      ${commonStyles};
      ${factoryStyles};
      ${commonMobileStyles};
      ${props =>
    (props.inheritedStyles && props.inheritedStyles[PRIMARY_THEME]
      ? props.inheritedStyles[PRIMARY_THEME][MOBILE]
      : '')};
      ${props =>
    (props.inheritedStyles && !props.inheritedStyles[PRIMARY_THEME]
      ? props.inheritedStyles[MOBILE]
      : '')};
      ${props =>
    (props.inheritedStyles &&
        !props.inheritedStyles[PRIMARY_THEME] &&
        !props.inheritedStyles[MOBILE]
      ? props.inheritedStyles
      : '')};
    `,
  },
};
