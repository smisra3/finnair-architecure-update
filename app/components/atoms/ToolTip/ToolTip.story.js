import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import ToolTip from './index';

storiesOf('ToolTip', module)
  .add('', () => (<ToolTip>Hello ToolTip</ToolTip>));
