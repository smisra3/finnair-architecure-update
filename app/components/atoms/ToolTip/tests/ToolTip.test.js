import { shallow, mount } from 'enzyme';
import { ToolTip } from '../ToolTip';
import { KEY_ENTER } from '../../../../constants';

/* eslint-disable import/prefer-default-export */
export const props = {
  children: <div>TEST</div>,
  className: 'string',
  label: 'TEST',
  container: 'tooltip',
};

const event = {
  which: KEY_ENTER,
};

describe('<ToolTip />', () => {
  let ToolTipComponent = '';
  beforeEach(() => {
    ToolTipComponent = shallow(<ToolTip {...props} />);
  });

  test('should render correctly', () => {
    expect(ToolTipComponent).toMatchSnapshot();
  });

  test('should set width when component is mounted', () => {
    const toolTipDiv = document.createElement('div');
    toolTipDiv.setAttribute('data-tool-tip', 'tooltip');
    document.body.appendChild(toolTipDiv);
    const mountedComponent = mount(<ToolTip {...props} />);
    expect(mountedComponent.length).toEqual(1);
  });

  test('handle events on Tooltip', () => {
    const handleKeyPress = jest.spyOn(ToolTipComponent.instance(), 'handleKeyPress');
    const handleMouseEnter = jest.spyOn(ToolTipComponent.instance(), 'handleMouseEnter');
    const handleMouseLeave = jest.spyOn(ToolTipComponent.instance(), 'handleMouseLeave');
    const handleClick = jest.spyOn(ToolTipComponent.instance(), 'handleClick');

    ToolTipComponent.simulate('mouseEnter');
    expect(handleMouseEnter).toHaveBeenCalled();
    expect(ToolTipComponent.state().isVisible).toBeTruthy();

    ToolTipComponent.simulate('keyPress', event);
    expect(handleKeyPress).toHaveBeenCalled();
    expect(ToolTipComponent.state().isVisible).toBeFalsy();

    ToolTipComponent.simulate('mouseLeave');
    expect(handleMouseLeave).toHaveBeenCalled();
    expect(ToolTipComponent.state().isVisible).toBeFalsy();

    ToolTipComponent.find('span[data-slnm-id="toolTipCTA"]').simulate('click');
    expect(handleClick).toHaveBeenCalled();
  });
});
