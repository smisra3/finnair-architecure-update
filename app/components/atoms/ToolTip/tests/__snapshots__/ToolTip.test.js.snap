// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToolTip /> should render correctly 1`] = `
<button
  className="string tooltip-parent  is-tooltip-hidden"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="tooltip-icon"
    data-slnm-id="toolTipCTA"
    onClick={[Function]}
    role="presentation"
  >
    <div>
      TEST
    </div>
  </span>
  <span
    className="tooltip-label"
    data-slnm-id="toolTipBubble"
    style={
      Object {
        "left": "-100%",
        "width": "13em",
      }
    }
  >
    <span
      className="tooltip-label-child"
    >
      TEST
    </span>
  </span>
</button>
`;
